src/AbstractValueParser.php

```diff
@@ -28,7 +41,9 @@ public function defaultsTo($defaultValue)
 
     public function required()
     {
-        if (is_null($this->value)) {
+        if ($this->invalid) {
+            throw $this->createInvalidValueException();
+        } elseif (is_null($this->value)) {
```

Comment:
I guess this code is a bit hard to read and should be rewritten to be more meaningful. I'll add a todo for this.